mcm/json_layer/request.py

```diff
@@ -3045,31 +3045,30 @@ def request_to_tasks(self, base, depend):
         if validation_info:
             # Average filter efficiency
             filter_efficiency_avg = sum([x['filter_efficiency'] for x in validation_info]) / len(validation_info)
-            filter_efficiency_threshold = 0.001
+            filter_efficiency_threshold = settings.get_value('filter_eff_threshold_for_thread_choice') #by legacy 0.001
+            cpu_efficiency_threshold = settings.get_value('cpu_efficiency_threshold_for_thread_choice')
+            best_efficiency_info = sorted(validation_info, key=lambda v: v['cpu_efficiency'])[-1]
```

Comment:
Assign these variables before the condition

```suggestion
            best_efficiency_info = sorted(validation_info, key=lambda v: v['cpu_efficiency'])[-1]
            filter_efficiency_margin = settings.get_value('filter_eff_margin_for_thread_choice') #0.05 # 5% margin
            max_efficiency = best_efficiency_info['cpu_efficiency']
```

Reply:
fixed with cdf609727562bd901b1b1f6bf48b56b8dc8d42ba